### PR TITLE
feat(podgrouper): Add JobSet Support to PodGrouper

### PR DESCRIPTION
**Pull request by rich7420**
## Description

1. This PR added PodGrouper plugin support for JobSet workloads so they can automatically create PodGroups for gang scheduling.
2. The plugin creates a separate PodGroup for each replicatedJob in a JobSet, which helps avoid sequencing deadlocks when using startupPolicyOrder: InOrder.
3. It calculates MinAvailable as replicas * min(parallelism, completions) to properly handle the replicas field and edge cases where completions differs from parallelism.
4. This enables users to take advantage of KAI Scheduler's gang scheduling for JobSet workloads without having to manually manage PodGroups.
If something wrong , plz correct me.

## Related Issues

Related #763 

## Checklist

> **Note:** Ensure your PR title follows the [Conventional Commits format](https://github.com/NVIDIA/KAI-Scheduler/blob/main/CONTRIBUTING.md#pr-title-guidelines) (e.g., `feat(scheduler): add new feature`)

- [x] Self-reviewed
- [x] Added/updated tests (if needed)
- [x] Updated documentation (if needed)

### CODE REVIEW

**Comment by rich7420** [comment]
cc @romanbaron , @davidLif

File: pkg/podgrouper/podgrouper/plugins/jobset/jobset_grouper.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	if !ok || replicatedJobName == "" {
+		return nil, fmt.Errorf("pod %s/%s missing required label %q", pod.Namespace, pod.Name, jobSetLabelReplicatedJobName)
+	}
+
+	pgMeta.Name = fmt.Sprintf(
+		"%s-%s-%s-%s",
+		jobSetPodGroupNamePrefix,
+		jobSetName,
+		replicatedJobName,
+		string(jobSetUID),
```
Comment (davidLif) [inline pkg/podgrouper/podgrouper/plugins/jobset/jobset_grouper.go]: Maybe a switch between the replicatedJobName and uid would look better? Then it's easier to understand that the uid belong to the jobset and not the replicatedJob.

**Comment by davidLif** [comment]
> 2. The plugin creates a separate PodGroup for each replicatedJob in a JobSet, which helps avoid sequencing deadlocks when using startupPolicyOrder: InOrder.

If the "Inorder" parameter isn't set, it would make more sence to schedule them as a single podgroup. You might want to test for the startupPolicyOrderspec and create the podgroups accordingly.


> 3. It calculates MinAvailable as replicas * min(parallelism, completions) to properly handle the replicas field and edge cases where completions differs from parallelism.

Please add an e2e test for the case of parallelism=2, completions=3, with the first pod completing 5 seconds earlier than the other 2.

**Review by davidLif — COMMENTED** [review]
Please add a few e2e tests under `"test/e2e/suites/integrations/third_party"`

**Comment by rich7420** [comment]
@davidLif thanks for the review!
I've updated it following your suggestion.

File: test_jobset.yaml [hunk 2]
```diff
@@ -0,0 +1,109 @@
+# Copyright 2025 NVIDIA CORPORATION
+# SPDX-License-Identifier: Apache-2.0
+
+# Test Scenario 1: Single ReplicatedJob with high parallelism
```
Comment (davidLif) [inline test_jobset.yaml]: This test example file should be in the main folder. I recommend changing it into an e2e test

File: pkg/podgrouper/podgrouper/plugins/jobset/jobset_grouper.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+	if !ok || replicatedJobName == "" {
+		return nil, fmt.Errorf("pod %s/%s missing required label %q", pod.Namespace, pod.Name, jobSetLabelReplicatedJobName)
+	}
+
+	startupPolicyOrder, err := getStartupPolicyOrder(topOwner)
+	if err != nil {
+		return nil, err
+	}
+
+	if startupPolicyOrder == "InOrder" {
```
Comment (davidLif) [inline pkg/podgrouper/podgrouper/plugins/jobset/jobset_grouper.go]: Please move the "InOrder" to const.

File: pkg/podgrouper/podgrouper/plugins/jobset/jobset_grouper.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+			jobSet.GetNamespace(), jobSet.GetName(), err)
+	}
+	if !found {
+		return "InOrder", nil
+	}
+	return order, nil
+}
+
+// getJobSetTotalMinAvailable calculates total minAvailable for all replicatedJobs.
+func getJobSetTotalMinAvailable(jobSet *unstructured.Unstructured) (int32, error) {
```
Comment (davidLif) [inline pkg/podgrouper/podgrouper/plugins/jobset/jobset_grouper.go]: The name of the function is too similar to "getJobSetMinAvailable". Maybe change them to something like "getReplicatedJobMinAvaliable" and "getJobSetMinAvailable"?

File: pkg/podgrouper/podgrouper/plugins/jobset/jobset_grouper.go [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+		if !ok {
+			continue
+		}
+
+		name, _, _ := unstructured.NestedString(rjMap, "name")
+		if name == "" {
+			continue
+		}
+
+		minAvailable, err := getJobSetMinAvailable(jobSet, name)
```
Comment (davidLif) [inline pkg/podgrouper/podgrouper/plugins/jobset/jobset_grouper.go]: Considering the fact that getJobSetMinAvailable iterates over all the replicatedJobs each time is called, maybe it would be better to extract the core per replicatedJob to a separate function and call it from getJobSetMinAvailable and getJobSetTotalMinAvailable

**Comment by rich7420** [comment]
@davidLif thanks for the review and suggestions!

File: pkg/podgrouper/podgrouper/plugins/jobset/jobset_grouper.go [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+
+		minAvailable, err := calculateReplicatedJobMinAvailable(rjMap, jobSet.GetNamespace(), jobSet.GetName(), name)
+		if err != nil {
+			return 0, err
+		}
+		totalMinAvailable += int64(minAvailable)
+	}
+
+	if totalMinAvailable <= 0 {
+		return 1, nil
```
Comment (davidLif) [inline pkg/podgrouper/podgrouper/plugins/jobset/jobset_grouper.go:224]: Why return 1 is the sum is 0? Maybe for some reason the user decided to "pause" the job and set the number of replicas to 0?

File: test/e2e/suites/integrations/third_party/jobset/jobset_test.go [hunk 7]
```diff
@@ -0,0 +1,558 @@
...
+	}
+)
+
+var _ = Describe("JobSet integration", Ordered, func() {
+	var (
+		testCtx *testcontext.TestContext
+	)
+
+	BeforeAll(func(ctx context.Context) {
+		testCtx = testcontext.GetConnectivity(ctx, Gomega)
```
Comment (davidLif) [inline test/e2e/suites/integrations/third_party/jobset/jobset_test.go]: Gomega cann't be passed as a parameter. Please look at other GetConnectivity calls.

File: test/e2e/suites/integrations/third_party/jobset/jobset_test.go [hunk 8]
```diff
@@ -0,0 +1,558 @@
...
+		parentQueue := queue.CreateQueueObject(utils.GenerateRandomK8sName(10), "")
+		childQueue := queue.CreateQueueObject(utils.GenerateRandomK8sName(10), parentQueue.Name)
+		testCtx.InitQueues([]*v2.Queue{childQueue, parentQueue})
+
+		skipIfJobSetNotInstalled(ctx, testCtx)
```
Comment (davidLif) [inline test/e2e/suites/integrations/third_party/jobset/jobset_test.go]: Please check if "JobSetNotInstalled" before creating the queues.

File: test/e2e/suites/integrations/third_party/jobset/jobset_test.go [hunk 9]
```diff
@@ -0,0 +1,558 @@
...
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// createJobSetWithDefaultParallelism creates a JobSet with single replica and default parallelism.
+func createJobSetWithDefaultParallelism(name, namespace, queueName string) *unstructured.Unstructured {
```
Comment (davidLif) [inline test/e2e/suites/integrations/third_party/jobset/jobset_test.go]: You can use a Unstructured like that, but importing the relevant struct would have also been fine.

File: test/e2e/suites/integrations/third_party/jobset/jobset_test.go [hunk 10]
```diff
@@ -0,0 +1,558 @@
...
+
+			// Wait for pods to be created
+			pods := waitForJobSetPods(ctx, testCtx, jobSetName, testNamespace, 3)
+
+			// Wait for all pods to be scheduled
+			wait.ForPodsScheduled(ctx, testCtx.ControllerClient, testNamespace, pods)
+
+			// Wait for first pod to complete (should complete 5 seconds earlier)
+			firstPod := pods[0]
+			Eventually(func() v1.PodPhase {
```
Comment (davidLif) [inline test/e2e/suites/integrations/third_party/jobset/jobset_test.go]: Please change `Eventually`"calls to the appropriate `wait` calls where possible. For example, here you can use `wait.ForPodSucceededOrError` .

Comment (davidLif) [reply test/e2e/suites/integrations/third_party/jobset/jobset_test.go]: See test/e2e/modules/wait

File: test/e2e/suites/integrations/third_party/jobset/jobset_test.go [hunk 11]
```diff
@@ -0,0 +1,558 @@
...
+			jobSetName := "inorder-jobset-" + utils.GenerateRandomK8sName(10)
+
+			jobSet := createJobSetWithStartupPolicy(jobSetName, testNamespace, testCtx.Queues[0].Name, "InOrder")
+			Expect(testCtx.ControllerClient.Create(ctx, jobSet)).To(Succeed())
+			defer func() {
+				Expect(testCtx.ControllerClient.Delete(ctx, jobSet)).To(Succeed())
+			}()
+
+			// Wait for pods and verify PodGroups are created separately
+			pods := waitForJobSetPods(ctx, testCtx, jobSetName, testNamespace, 2)
```
Comment (davidLif) [inline test/e2e/suites/integrations/third_party/jobset/jobset_test.go]: How does this call validates that 2 separate podgroups have been created?

File: test/e2e/suites/integrations/third_party/jobset/jobset_test.go [hunk 12]
```diff
@@ -0,0 +1,558 @@
...
+			jobSetName := "anyorder-jobset-" + utils.GenerateRandomK8sName(10)
+
+			jobSet := createJobSetWithStartupPolicy(jobSetName, testNamespace, testCtx.Queues[0].Name, "AnyOrder")
+			Expect(testCtx.ControllerClient.Create(ctx, jobSet)).To(Succeed())
+			defer func() {
+				Expect(testCtx.ControllerClient.Delete(ctx, jobSet)).To(Succeed())
+			}()
+
+			// Wait for pods and verify they all use the same PodGroup
+			pods := waitForJobSetPods(ctx, testCtx, jobSetName, testNamespace, 2)
```
Comment (davidLif) [inline test/e2e/suites/integrations/third_party/jobset/jobset_test.go]: How does this call validates that only one podgroup have been created?

File: test/e2e/suites/integrations/third_party/jobset/jobset_test.go [hunk 13]
```diff
@@ -0,0 +1,558 @@
...
+	AfterEach(func(ctx context.Context) {
+		testCtx.TestContextCleanup(ctx)
+	})
+
+	AfterAll(func(ctx context.Context) {
+		testCtx.ClusterCleanup(ctx)
+	})
+
+	Context("JobSet submission", func() {
+		It("should schedule pods with parallelism=2 and completions=3, first pod completes 5 seconds earlier", func(ctx context.Context) {
```
Comment (davidLif) [inline test/e2e/suites/integrations/third_party/jobset/jobset_test.go]: This is a good test, please set default-staleness-grace-period to 1 second at the begining of the test and remove this config at the end. Please look at `test/e2e/modules/configurations/feature_flags` before implementing.

File: test/e2e/suites/integrations/third_party/jobset/jobset_test.go [hunk 14]
```diff
@@ -0,0 +1,558 @@
...
+	crdList, err := testCtx.KubeClientset.ApiextensionsV1().CustomResourceDefinitions().List(ctx, metav1.ListOptions{
+		FieldSelector: "metadata.name=jobsets.jobset.x-k8s.io",
+	})
+	Expect(err).NotTo(HaveOccurred())
+	if len(crdList.Items) == 0 {
+		Skip("JobSet CRD not installed, skipping JobSet integration tests")
+	}
+}
+
+func createJobSet(name, namespace, queueName string, parallelism, completions int32) *unstructured.Unstructured {
```
Comment (davidLif) [inline test/e2e/suites/integrations/third_party/jobset/jobset_test.go]: Please import the relevant struct from the JobSet repo instead of manipulating Unstructured for all JobSet creations.

File: test/e2e/suites/integrations/third_party/jobset/jobset_test.go [hunk 15]
```diff
@@ -0,0 +1,558 @@
...
+				g.Expect(len(podGroups.Items)).To(Equal(1))
+				// MinMember should default to 1 (replicas=1, parallelism defaults to 1)
+				g.Expect(podGroups.Items[0].Spec.MinMember).To(Equal(int32(1)))
+			}, time.Minute).Should(Succeed())
+		})
+	})
+})
+
+func skipIfJobSetNotInstalled(ctx context.Context, testCtx *testcontext.TestContext) {
+	crdList, err := testCtx.KubeClientset.ApiextensionsV1().CustomResourceDefinitions().List(ctx, metav1.ListOptions{
```
Comment (davidLif) [inline test/e2e/suites/integrations/third_party/jobset/jobset_test.go]: Please use `crd.SkipIfCrdIsNotInstalled` instead. See `test/e2e/suites/integrations/third_party/kubeflow.org/pytorch/pytorch_test.go` for example

**Review by davidLif — COMMENTED** [review]
1- Please add a `deploy_jobset.sh` script under `hack/third_party_integrations` to enable installation for `run-e2e-kind.sh` (and add a call to this script appropriatly). 

2- After adding deploy_jobset.sh`, please validate that all Jobset e2e tests pass on a kinf cluster deployed by `run-e2e-kind.sh --test-third-party-integrations` (You can use --preserve-cluster if you need to debug some of the tests)

@rich7420  Thank you very much for you patience and effort!

File: pkg/podgrouper/podgrouper/plugins/jobset/jobset_grouper.go (hunk 6 above)
Comment (rich7420) [reply pkg/podgrouper/podgrouper/plugins/jobset/jobset_grouper.go:224]: I set it to 1 because the PodGroup CRD validation requires minMember to be at least 1. However, you are right regarding the paused job scenario (replicas=0). Forcing it to 1 might cause the PodGroup to hang waiting for a pod that will never exist.
Perhaps we should skip creating the PodGroup entirely if totalMinAvailable calculates to 0? WDYT?

Comment (davidLif) [reply pkg/podgrouper/podgrouper/plugins/jobset/jobset_grouper.go:224]: The the PodGroup CRD validation requires minMember > 1, then I think the code you wrote (minMember 1) is the best option.
I wouldn't try to "skip creating the PodGroup". 
It would create a mess for cases like this:  if the podgroup existed, the use changed the JobSet, the podgroup would be deleted, the use changed the JobSet again, and now we create a new podgroup for the same entity again.

**Comment by rich7420** [comment]
@davidLif thanks for the patient reviewing! I've updated.

File: hack/third_party_integrations/deploy_jobset.sh [hunk 16]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+# Copyright 2025 NVIDIA CORPORATION
+# SPDX-License-Identifier: Apache-2.0
+set -e
+
+kubectl apply --server-side -k "github.com/kubernetes-sigs/jobset.git/config/default?ref=v0.2.0"
```
Comment (davidLif) [inline hack/third_party_integrations/deploy_jobset.sh]: When i tried to run it on my machine, the image pull failed. 
I recommend using `kubectl apply --server-side -f https://github.com/kubernetes-sigs/jobset/releases/download/v0.10.1/manifests.yaml` instead, as described by https://github.com/kubernetes-sigs/jobset?tab=readme-ov-file#installation .

**Comment by rich7420** [comment]
@davidLif thanks! I've updated.

**Review by davidLif — APPROVED** [review]

**Comment by rich7420** [comment]
thanks @davidLif , @gshaibi for the review and patience !
What’s the next step I can help with?
